### PR TITLE
Add simple ODE demo

### LINKED ISSUE #100: Add ODE demo
It would be nice to have a demo illustrating solving an ODE, i.e., in the Real function space.

I was thinking of just solving

$\frac{du}{dt}=u,\quad u(t)=1,\quad t\in[0,1],$

(which has analytical solution $u(t)=e^t$) using Forward Euler, Backward Euler, and Crank-Nicolson and comparing the results.

[Motivation: a simple example like this would be a good springboard for trying out parallel-in-time methods in Goalie.]

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #100.

Also introduces a new `PointSeq` class for use with ODE problems.

### CODE REVIEW

**Comment by jwallwork23** [comment]
Test suite seems to be failing for the same reason currently under investigation in #104.

**Comment by jwallwork23** [comment]
@ddundo Since you mentioned you'd already been looking at this, would you be happy to review it?

**Review by ddundo — APPROVED** [review]
Looks nice @jwallwork23! Left a few small comments :)

File: goalie/point_seq.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        r"""
+        :arg time_partition: the :class:`~.TimePartition` which partitions the temporal
+            domain
+        :kwarg get_function_spaces: a function, whose only argument is a
+            :class:`~.MeshSeq`, which constructs prognostic
+            :class:`firedrake.functionspaceimpl.FunctionSpace`\s for each subinterval
+        :kwarg get_initial_condition: a function, whose only argument is a
+            :class:`~.MeshSeq`, which specifies initial conditions on the first mesh
+        :kwarg get_form: a function, whose only argument is a :class:`~.MeshSeq`, which
+            returns a function that generates the PDE weak form
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply goalie/point_seq.py]: Good spot - this was a copy-paste error. Addressed in e4a9708.

File: demos/ode.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    get_solver=get_solver,
+)
+
+# We can solve the ODE using the :meth:`~.MeshSeq.solve_forward` method and extract the
+# solution trajectory as follows. Note that the trajectory is prepended with the initial
+# value. ::
+solutions = point_seq.solve_forward()["u"]["forward"]
+forward_euler_trajectory = [1] + [
+    sol.dat.data[0] for by_sub in solutions for sol in by_sub
+]
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply demos/ode.py]: Good points! Addressed in 3e0b6a4e24ee144cdc582328287b4120dec2ab0d and 32a804984fb4aace428c1d6967be02e50aeb4c93.

File: demos/ode.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+#
+# Goalie was designed primarily with partial differential equations (PDEs) in mind.
+# However, it can also be used to solve ordinary differential equations (ODEs).
+#
+# For example, consider the scalar, linear ODE,
+#
+# .. math::
+#    \frac{\mathrm{d}u}{\mathrm{d}t} = u,\quad u(0) = 1,
+#
+# which we solve for :math:`u(t)` over the time period :math:`t\in[0,1]`. It is
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply demos/ode.py]: Fixed in 002900e.

File: demos/ode.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+# explicit function of solution approximation at time level :math:`i`.
+#
+# .. math::
+#    \frac{u_{i+1} - u_i}{\Delta t} = u_i,
+#
+# where :math:`u_k` denotes the approximation at time level `k` and :math:`\Delta k` is
+# the timestep length. This expression may be rearranged as
+#
+# .. math::
+#    u_{i+1} - u_i + - \Delta t u_i = 0.
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply demos/ode.py]: Good spot. Fixed in 323c0c3.

File: demos/ode.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+#
+# Even though there are no spatial derivatives in our problem, we still have to
+# represent the problem in a finite element formulation in order to solve it using
+# Goalie. Recall that functions in our function space, :math:`R` are just scalar valued.
+# To use the finite element notation, we multiply both sides of the equation by a test
+# function and integrate:
+#
+# .. math::
+#    \int (u_{i+1} - u_i  - \Delta t u_i) v \mathrm{d}R = 0, \forall v\in R.
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply demos/ode.py]: Fixed in aaadddd, also adding integration bounds.

File: demos/ode.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+# Implicit Euler because the approximation at the next time level is represented as an
+# implicit function of the approximation at the current time level:
+#
+# .. math::
+#    \frac{u_{i+1} - u_i}{\Delta t} = u_{i+1}.
+#
+# Similarly to the above, this gives rise to
+#
+# .. math::
+#    \int (u_{i+1} - u_{i} - \Delta t u_{i+1}) v \mathrm{d}R, \forall v\in R.
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply demos/ode.py]: Fixed in aaadddd, also adding integration bounds.

File: demos/ode.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    get_function_spaces=get_function_spaces,
+    get_initial_condition=get_initial_condition,
+    get_form=get_form_backward_euler,
+    get_solver=get_solver,
+)
+solutions = point_seq.solve_forward()["u"]["forward"]
+
+backward_euler_trajectory = [1] + [
+    sol.dat.data[0] for by_sub in solutions for sol in by_sub
+]
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply demos/ode.py]: Addressed in 3e0b6a4e24ee144cdc582328287b4120dec2ab0d.

**Comment by jwallwork23** [comment]
Thanks for the review @ddundo. I think I addressed all your concerns. Also used `float(sol)` instead of `sol.dat.data[0]` for simplicity and to avoid exposing underlying Firedrake data structures in an early demo.

**Comment by ddundo** [comment]
Thanks @jwallwork23, I think this is much clearer now!

**Pull request merged**